examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -20,56 +20,63 @@
+import org.apache.kafka.common.errors.UnsupportedVersionException;
 
-import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
 import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.atomic.AtomicLong;
+
+import static java.time.Duration.ofMillis;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singleton;
 
 /**
- * A demo class for how to write a customized EOS app. It takes a consume-process-produce loop.
- * Important configurations and APIs are commented.
+ * This class implements a read-process-write application.
  */
-public class ExactlyOnceMessageProcessor extends Thread {
-
-    private static final boolean READ_COMMITTED = true;
-
+public class ExactlyOnceMessageProcessor extends Thread implements ConsumerRebalanceListener {
+    private final String bootstrapServers;
     private final String inputTopic;
     private final String outputTopic;
-    private final String transactionalId;
     private final String groupInstanceId;
+    private final CountDownLatch latch;
+    private final String transactionalId;
+    private volatile boolean closed;
 
     private final KafkaProducer<Integer, String> producer;
     private final KafkaConsumer<Integer, String> consumer;
 
-    private final CountDownLatch latch;
-
-    public ExactlyOnceMessageProcessor(final String inputTopic,
-                                       final String outputTopic,
-                                       final int instanceIdx,
-                                       final CountDownLatch latch) {
+    public ExactlyOnceMessageProcessor(String threadName,
+                                       String bootstrapServers,
+                                       String inputTopic,
+                                       String outputTopic,
+                                       CountDownLatch latch) {
+        super(threadName);
+        this.bootstrapServers = bootstrapServers;
         this.inputTopic = inputTopic;
         this.outputTopic = outputTopic;
-        this.transactionalId = "Processor-" + instanceIdx;
+        this.transactionalId = threadName;
         // It is recommended to have a relatively short txn timeout in order to clear pending offsets faster.
         final int transactionTimeoutMs = 10000;
         // A unique transactional.id must be provided in order to properly use EOS.
         producer = new Producer(outputTopic, true, transactionalId, true, -1, transactionTimeoutMs, null).get();
         // Consumer must be in read_committed mode, which means it won't be able to read uncommitted data.
         // Consumer could optionally configure groupInstanceId to avoid unnecessary rebalances.
-        this.groupInstanceId = "Txn-consumer-" + instanceIdx;
+        this.groupInstanceId = threadName;
```

Comment:
I think it makes sense to keep different name for group instance id and transactional id. Could we set different value for them?